### PR TITLE
Use a fixed-size daemon thread pool for the Kafka Binder

### LINKED ISSUE #500: Applications using the Kafka binder do not shut down on context close()
When the context is closed (e.g. via actuator endpoint), Reactor keeps running on the thread pool of `QueueingMesageListenerInvoker`, preventing the application to shutdown. This is related to https://jira.spring.io/browse/INTEXT-214.

### PR DESCRIPTION
**Pull request by mbogoevici**
Fix #500

- Initialize a fixed-size ExecutorService of the size of the concurrency setting of the binder
- Clean up the ExecutorService on unbind
- Remove unused `spyOn` support

### CODE REVIEW

File: spring-cloud-stream-binders/spring-cloud-stream-binder-kafka/src/main/java/org/springframework/cloud/stream/binder/kafka/KafkaMessageChannelBinder.java [hunk 1]
```diff
@@ -460,9 +465,28 @@ public Object doWithRetry(RetryContext context) throws RuntimeException {
 		final FixedSubscriberChannel bridge = new FixedSubscriberChannel(rh);
 		bridge.setBeanName("bridge." + name);
 
-		final KafkaMessageListenerContainer messageListenerContainer =
-				createMessageListenerContainer(properties, group, null, listenedPartitions, referencePoint);
+		Assert.isTrue(StringUtils.hasText(null) ^ !CollectionUtils.isEmpty(listenedPartitions),
```
Comment (markfisher) [inline spring-cloud-stream-binders/spring-cloud-stream-binder-kafka/src/main/java/org/springframework/cloud/stream/binder/kafka/KafkaMessageChannelBinder.java]: is this `StringUtils.hasText(null)` intentional?

File: spring-cloud-stream-binders/spring-cloud-stream-binder-kafka/src/main/java/org/springframework/cloud/stream/binder/kafka/KafkaMessageChannelBinder.java [hunk 2]
```diff
@@ -460,9 +465,28 @@ public Object doWithRetry(RetryContext context) throws RuntimeException {
...
+				"Exactly one of topic or a list of listened partitions must be provided");
+		final KafkaMessageListenerContainer messageListenerContainer = new KafkaMessageListenerContainer(connectionFactory,
+				listenedPartitions.toArray(new Partition[listenedPartitions.size()]));
+
+		if (logger.isDebugEnabled()) {
+			logger.debug("Listening to topic " + null);
```
Comment (markfisher) [inline spring-cloud-stream-binders/spring-cloud-stream-binder-kafka/src/main/java/org/springframework/cloud/stream/binder/kafka/KafkaMessageChannelBinder.java]: is `null` intentional here also?

File: spring-cloud-stream-binders/spring-cloud-stream-binder-kafka/src/main/java/org/springframework/cloud/stream/binder/kafka/KafkaMessageChannelBinder.java (hunk 1 above)
Comment (mbogoevici) [inline spring-cloud-stream-binders/spring-cloud-stream-binder-kafka/src/main/java/org/springframework/cloud/stream/binder/kafka/KafkaMessageChannelBinder.java]: Good point. This is an artifact of inlining the old code. Initially we had support for passing both a topic or a list of partitions (for the never used `spyOn`),  so now it's always null - will fix.

**Comment by markfisher** [comment]
rebased, squashed, and merged as https://github.com/spring-cloud/spring-cloud-stream/commit/0bd7899a77977cad29fbdd67e1a908b552c77e32

**Pull request closed without merging**